upbc/protoc-gen-upb.cc

```diff
@@ -369,6 +329,346 @@ std::string CTypeConst(const protobuf::FieldDescriptor* field) {
+  // descriptor must be from this layout's file.
+  upb_MiniTable* GetMiniTable(const protobuf::Descriptor* m) const;
+  const upb_MiniTable_Extension* GetExtension(
+      const protobuf::FieldDescriptor* fd) const;
+
+  // Get the initializer for the given sub-message/sub-enum link.
+  static std::string GetSub(upb_MiniTable_Sub sub);
+
+ private:
+  // Functions to build mini-tables for this file's messages and extensions.
+  void BuildMiniTables(const protobuf::FileDescriptor* fd);
+  void BuildExtensions(const protobuf::FileDescriptor* fd);
+  upb_MiniTable* MakeMiniTable(const protobuf::Descriptor* m);
+  upb_MiniTable* MakeRegularMiniTable(const protobuf::Descriptor* m);
+  uint64_t GetMessageModifiers(const protobuf::Descriptor* m);
+  uint64_t GetFieldModifiers(const protobuf::FieldDescriptor* f);
+  void ResolveIntraFileReferences();
+
+  // When we are generating code, tables are linked to sub-tables via name (ie.
+  // a string) rather than by pointer.  We need to emit an initializer like
+  // `&foo_sub_table`.  To do this, we store `const char*` strings in all the
+  // links that would normally be pointers:
+  //    field -> sub-message
+  //    field -> enum table (proto2 only)
+  //    extension -> extendee
+  //
+  // This requires a bit of reinterpret_cast<>(), but it's confined to a few
+  // functions.  We tag the pointer so we know which member of the union to
+  // initialize.
+  enum SubTag {
+    kNull = 0,
+    kMessage = 1,
+    kEnum = 2,
+    kMask = 3,
+  };
+
+  static upb_MiniTable_Sub PackSub(const char* data, SubTag tag);
+  static bool IsNull(upb_MiniTable_Sub sub);
+  void SetSubTableStrings();
+  upb_MiniTable_Sub PackSubForField(const protobuf::FieldDescriptor* f,
+                                    const upb_MiniTable_Field* mt_f);
+  const char* AllocStr(const std::string& str);
+
+ private:
+  typedef absl::flat_hash_map<const protobuf::Descriptor*, upb_MiniTable*>
+      TableMap;
+  typedef absl::flat_hash_map<const protobuf::FieldDescriptor*,
+                              upb_MiniTable_Extension>
+      ExtensionMap;
+  upb::Arena arena_;
+  TableMap table_map_;
+  ExtensionMap extension_map_;
+  upb_MiniTablePlatform platform_;
+};
+
+upb_MiniTable* FilePlatformLayout::GetMiniTable(
+    const protobuf::Descriptor* m) const {
+  auto it = table_map_.find(m);
+  assert(it != table_map_.end());
+  return it->second;
+}
+
+const upb_MiniTable_Extension* FilePlatformLayout::GetExtension(
+    const protobuf::FieldDescriptor* fd) const {
+  auto it = extension_map_.find(fd);
+  assert(it != extension_map_.end());
+  return &it->second;
+}
+
+void FilePlatformLayout::ResolveIntraFileReferences() {
+  // This properly resolves references within a file, in order to set any
+  // necessary flags (eg. is a map).
+  for (const auto& pair : table_map_) {
+    upb_MiniTable* mt = pair.second;
+    // First we properly resolve for defs within the file.
+    for (const auto& f : FieldNumberOrder(pair.first)) {
```

Comment:
looks like you are taking a `&` to a `*`.  If you can make this a `const auto*` that would be cleaner

Reply:
Done.